marker_api/src/ast/item.rs

```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    // * The rustc model makes sense, so have a kind with a restriction
+    // * have according methods, that check for `pub(crate)`
+    // * document possible equivalency between `pub(crate)` and `pub(super)`
+    // * Document that things can be reexported
+    // * Test how visibility only to lower scopes works `pub(self::sub_module)`
 }
 
-impl<'ast> Debug for Visibility<'ast> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Visibility {{ /* WIP: See rust-marker/marker#26 */}}")
-            .finish()
+impl<'ast> Visibility<'ast> {
+    pub fn span(&self) -> Option<&Span<'ast>> {
+        self.span.copy().map(|span| with_cx(self, |cx| cx.span(span)))
     }
-}
 
-#[cfg(feature = "driver-api")]
-impl<'ast> Visibility<'ast> {
-    pub fn new(item_id: ItemId) -> Self {
-        Self {
-            _lifetime: PhantomData,
-            _item_id: item_id,
+    /// Returns `true` if the item is declared as public, without any restrictions.
+    ///
+    /// ```
+    /// // This returns `true`
+    /// pub fn unicorn() {}
+    ///
+    /// // This returns `false`, since the visibility is restricted to a specified path.
+    /// pub(crate) fn giraffe() {}
+    ///
+    /// // This returns `false`, since the visibility is not defined
+    /// fn dragon() {}
+    /// ```
+    ///
+    /// See [`Visibility::is_pub_with_path`] to detect pub declarations with a
+    /// defined path.
+    pub fn is_pub(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Public | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns `true` if the item is declared as `pub(...)` with a path in brackets
+    /// that defines the scope, where the item is visible.
+    pub fn is_pub_with_path(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Path(_) | VisibilityKind::Crate(_))
+    }
+
+    /// Returns `true` if the visibility is declared as `pub(crate)`. This is a
+    /// special case of the `pub(<path>)` visibility.
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    ///
+    /// ```rs
+    /// // lib.rs
+    ///
+    /// mod example_1 {
+    ///     // Returns `false` since no visibility is declared
+    ///     fn foo() {}
+    ///
+    ///     // Returns `false` since the item is not visible from the root of the crate.
+    ///     pub(in crate::example_1) fn bar() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to the root of the crate.
+    ///     pub(crate) baz() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to `super` which is the
+    ///     // root of the crate.
+    ///     pub(crate) boo() {}
+    /// }
+    ///
+    /// // Returns `false` since the visibility is not restricted
+    /// fn example_in_root() {}
```

Comment:
This makes it again special case. I don't like how syntactic and semantic meaning of `pub(crate)` are intermixed here.
What I would expect is smth like this:
- `ast::Visibility::is_pub_crate()`: checks if the visibility is spelled exactly as `pub(crate)` returning false for any other syntax

- `ast::Visibility::is_pub_in_path()`: checks if the visibility is spelled exactly as `pub(in ...)` returning false for any other syntax

- `sem::Visibility::is_pub_crate()`: checks if the visibility makes the symbol accessible in the crate root (`pub(crate)`, `pub(super)`, `pub(in ...)`, default visibility in `lib.rs` all would return `true`)

- `sem::Visibility::is_pub_in_path(path)`: checks if the visibility makes the symbol accessible under the given path (any syntax e.g. `pub(crate)`, `pub(in ...)` may return `true` only if the resulting resolved visibility allows the symbol to be accessed in `path` module).

Reply:
See: https://github.com/rust-marker/marker/pull/294#discussion_r1375321200